### PR TITLE
fix: ensure correct alter table alter column statement is generated on data type changes in alembic

### LINKED ISSUE #471: Changing Column data types fails in Alembic migration due to missing SET DATA TYPE
#### Environment details

  - OS type and version: Ubuntu 20.04
  - Python version: 3.9.13
  - pip version: 21.3.1
  - `sqlalchemy-bigquery` version: 1.4.4
  - `sqlalchemy` version: 1.4.27
  - `alembic` version: 1.8.0

#### Steps to reproduce

  1. Create a simple table class with single column with data type 'String'
  2. Autogenerate a migration to deploy the table with single column to BigQuery
  3. Deploy the table to BigQuery using `alembic upgrade head`
  4. Ensure alembic's env.py `run_migrations_online` passes the parameter `compare_type=True` to the `context.configure()` call in that function.
  5. Change the data type of the single column from `String` to `ARRAY(String)`
  6. Autogenerate a new migration and check that it has picked up the data type change.
  7. Attempt to deploy the second migration to BigQuery using `alembic upgrade head`
  8. Observe that sqlalchemy throws a `DatabaseError`: 

> sqlalchemy.exc.DatabaseError: (google.cloud.bigquery.dbapi.exceptions.DatabaseError) 400 Syntax error: Expected keyword DROP or keyword SET but got keyword TYPE at [1:72]

#### Expected behaviour

- ~~The migration executes properly~~
- The migration executes properly in case the type conversion is allowed according to the table [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/conversion_rules#comparison_chart), otherwise it throws an error that the type conversion is not allowed (see [comment below](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/471#issuecomment-1464047819))
- the ALTER TABLE statement uses the syntax described here:

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#alter_column_set_data_type_statement

#### Code example

```python
# ORM model first migration (my_app/models/orm_model.py)
from sqlalchemy.orm import declarative_base

from sqlalchemy import Column, String

BQBase = declarative_base()

class SomeTable(BQBase):
    __tablename__ = "some_table"

    some_column = Column(
        String,
        name="some_column",
        nullable=False
    )

# ORM model second migration
from sqlalchemy.orm import declarative_base

from sqlalchemy import Column, String, ARRAY

BQBase = declarative_base()

class SomeTable(DWHBase):
    __tablename__ = "some_table"

    some_column = Column(
        ARRAY(String),
        name="some_column",
        nullable=False
    )

# alembic/env.py

import os
from logging.config import fileConfig

from sqlalchemy import engine_from_config
from sqlalchemy import pool

from dotenv import load_dotenv

from alembic import context
from my_app.models.orm_model import BQBase

# load environment variables
load_dotenv()

# this is the Alembic Config object, which provides
# access to the values within the .ini file in use.
config = context.config

# Interpret the config file for Python logging.
# This line sets up loggers basically.
fileConfig(config.config_file_name)

# add your model's MetaData object here
# for 'autogenerate' support
# from myapp import mymodel
# target_metadata = mymodel.Base.metadata
target_metadata = BQBase.metadata

# other values from the config, defined by the needs of env.py,
# can be acquired:
# my_important_option = config.get_main_option("my_important_option")
# ... etc.


def run_migrations_offline():
    """Run migrations in 'offline' mode.

    This configures the context with just a URL
    and not an Engine, though an Engine is acceptable
    here as well.  By skipping the Engine creation
    we don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the
    script output.

    """
    url = os.getenv("BIGQUERY_CONNECTION_URL")
    context.configure(
        url=url,
        target_metadata=target_metadata,
        compare_type=True,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
    )

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online():
    """Run migrations in 'online' mode.

    In this scenario we need to create an Engine
    and associate a connection with the context.

    """
    connectable = engine_from_config(
        {"sqlalchemy.url": os.getenv("BIGQUERY_CONNECTION_URL")},
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            compare_type=True
        )

        with context.begin_transaction():
            context.run_migrations()


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()

# first migration
"""Add schema

Revision ID: 1
Revises: 
Create Date: 

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '1'
down_revision = None
branch_labels = None
depends_on = None

def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('some_table',
    sa.Column('id', sa.String(), nullable=False),
    )

def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('some_table')

# second migration

"""Change data type of some_table.some_column

Revision ID: 2
Revises: 1
Create Date: 

"""
from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision = '2'
down_revision = '1'
branch_labels = None
depends_on = None

def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.alter_column('some_table', 'some_column',
               existing_type=sa.String(),
               type_=sa.ARRAY(sa.String()),
               existing_nullable=False)
    # ### end Alembic commands ###

def downgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.alter_column('some_table', 'some_column',
               existing_type=sa.ARRAY(sa.String()),
               type_=sa.String(),
               existing_nullable=False)
    # ### end Alembic commands ###

```

#### Stack trace
```
Traceback (most recent call last):
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/bin/alembic", line 8, in <module>
    sys.exit(main())
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/config.py", line 590, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/config.py", line 584, in main
    self.run_cmd(cfg, options)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/config.py", line 561, in run_cmd
    fn(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/command.py", line 322, in upgrade
    script.run_env()
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "./my_project/alembic/env.py", line 87, in <module>
    run_migrations_online()
  File "./my_project/alembic/env.py", line 81, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/runtime/environment.py", line 853, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/runtime/migration.py", line 623, in run_migrations
    step.migration_fn(**kw)
  File "/home/someUser/Documents/my_project/alembic/versions/32ccb975b7eb_change_type_of_education_institution_.py", line 21, in upgrade
    op.alter_column('some_table', 'some_column',
  File "<string>", line 8, in alter_column
  File "<string>", line 3, in alter_column
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/operations/ops.py", line 1880, in alter_column
    return operations.invoke(alt)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/operations/base.py", line 394, in invoke
    return fn(self, operation)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/operations/toimpl.py", line 50, in alter_column
    operations.impl.alter_column(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/ddl/impl.py", line 275, in alter_column
    self._exec(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/alembic/ddl/impl.py", line 195, in _exec
    return conn.execute(construct, multiparams)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1289, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/sql/ddl.py", line 77, in _execute_on_connection
    return connection._execute_ddl(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1381, in _execute_ddl
    ret = self._execute_context(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1845, in _execute_context
    self._handle_dbapi_exception(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2026, in _handle_dbapi_exception
    util.raise_(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1802, in _execute_context
    self.dialect.do_execute(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 719, in do_execute
    cursor.execute(statement, parameters)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/_helpers.py", line 489, in with_closed_check
    return method(self, *args, **kwargs)
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 166, in execute
    self._execute(
  File "/home/someUser/.cache/pypoetry/virtualenvs/my-project-py3.9/lib/python3.9/site-packages/google/cloud/bigquery/dbapi/cursor.py", line 205, in _execute
    raise exceptions.DatabaseError(exc)
sqlalchemy.exc.DatabaseError: (google.cloud.bigquery.dbapi.exceptions.DatabaseError) 400 Syntax error: Expected keyword DROP or keyword SET but got keyword TYPE at [1:72]

(job ID: )

                               -----Query Job SQL Follows-----                                

    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |
   1:ALTER TABLE `some_table` ALTER COLUMN `some_column` TYPE ARRAY<STRING>
    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |    .    |
[SQL: ALTER TABLE `some_table` ALTER COLUMN `some_column` TYPE ARRAY<STRING>]
(Background on this error at: https://sqlalche.me/e/14/4xp6)

```

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

### PR DESCRIPTION
**Pull request by benvdh-incentro**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #471  🦕

### CODE REVIEW

**Comment by benvdh-incentro** [comment]
@chalmerlowe it turns out the alembic system test indeed failed because I moved the code to a separate module. I added it back to `base.py` and system test suite for 3.8 ran successfully again.

I've also made an attempt to add an additional system test, that tests altering column types... Somehow I only got that part working with a column set to nullable.

In case of columns that are not NULLABLE (or mode required) - as it is the case with the transactions table in the alembic system test - it failed with the following error:

```
sqlalchemy.exc.DatabaseError: (google.cloud.bigquery.dbapi.exceptions.DatabaseError) 400 Column `account` cannot change mode via SET DATA TYPE.

Location: US
Job ID: a1783672-fe58-478b-aaae-f86a06ba83f3

[SQL: ALTER TABLE `transactions` ALTER COLUMN `account` SET DATA TYPE NUMERIC]
(Background on this error at: http://sqlalche.me/e/13/4xp6)
```

I have tried playing around with the `nullable` and `existing_nullable` arguments of `alter_column`, but no luck so far. Perhaps it has something to do with the fact that in bigquery a column with mode `REQUIRED` (non-nullable), can be changed to `NULLABLE` but not the other way around, while the visit_column_nullable() method of Alembic's `DefaultImpl` assumes you can convert from non-nullable to nullable, and vice versa.

Perhaps you can have a look at this one, and re-run the CI?

**Comment by benvdh-incentro** [comment]
@chalmerlowe I got the coverage error as well locally, I am guessing it's because the ImportError in `base.py` is never reached, as alembic is usually available... but I have no idea how we could test for such an error, nor how relevant it is in actually solving the problem at hand...

**Comment by chalmerlowe** [comment]
based on the results in coverage, the statement that is not tested appears to be the result of our function as seen in line `1071`:

```
Name                                     Stmts   Miss Branch BrPart  Cover   Missing
------------------------------------------------------------------------------------
sqlalchemy_bigquery/__init__.py             14      0      0      0   100%
sqlalchemy_bigquery/_helpers.py             43      0     14      0   100%
sqlalchemy_bigquery/_struct.py              47      0     16      0   100%
sqlalchemy_bigquery/_types.py               57      0     18      0   100%
sqlalchemy_bigquery/base.py                487      1    144      0    99%   1071
```

This snippet:
```
return "%s %s %s" % (
            alter_table(compiler, element.table_name, element.schema),
            alter_column(compiler, element.column_name),
            "SET DATA TYPE %s" % format_type(compiler, element.type_),
        )
        ```

I have not messed with coverage much, I am not sure how coverage goes about calculating "coverage" via test.
If they predominantly look for tests that match based on file name and based on test function name?

I can poke around to see what I can find out, but won't be til later today or tomorrow, prolly.

File: sqlalchemy_bigquery/base.py [hunk 1]
```diff
@@ -1059,6 +1061,19 @@ def __init__(self, *args, **kwargs):
...
 
     class SqlalchemyBigqueryImpl(impl.DefaultImpl):
         __dialect__ = "bigquery"
+
+    @compiles(ColumnType, "bigquery")
+    def visit_column_type(element: ColumnType, compiler: DDLCompiler, **kw) -> str:
+        """Replaces the visit_column_type() function in alembic/alembic/ddl/base.py.
+        
+        The alembic version ends in TYPE <element type>, but bigquery requires this syntax:
+        SET DATA TYPE <element type>"""
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:1074]:
```suggestion
        SET DATA TYPE <element type>"""
        
```

File: sqlalchemy_bigquery/base.py [hunk 2]
```diff
@@ -1071,6 +1073,20 @@ def __init__(self, *args, **kwargs):
...
     from alembic.ddl import impl
+    from alembic.ddl.base import ColumnType, format_type, alter_table, alter_column
 
     class SqlalchemyBigqueryImpl(impl.DefaultImpl):
         __dialect__ = "bigquery"
+
+    @compiles(ColumnType, "bigquery")
+    def visit_column_type(element: ColumnType, compiler: DDLCompiler, **kw) -> str:
+        """Replaces the visit_column_type() function in alembic/alembic/ddl/base.py.
+
```
Comment (chalmerlowe) [inline sqlalchemy_bigquery/base.py:1084]:
```suggestion

```

**Review by chalmerlowe — APPROVED** [review]

**Comment by benvdh-incentro** [comment]
@chalmerlowe thank you too for the last bit to get the code through the pipeline. It was my pleasure to help and learn a lot along the way.

**Pull request merged**